File: Python/gabaritos/gabarito26.py
import time

def busca_linear(lst, alvo):
    for i, item in enumerate(lst):
        if item == alvo:
            return i
    return -1

def busca_binaria(lst, alvo):
    left, right = 0, len(lst) - 1
    while left <= right:
        mid = (left + right) // 2
        if lst[mid] == alvo:
            return mid
        elif lst[mid] < alvo:
            left = mid + 1
        else:
            right = mid - 1
    return -1

lista = list(range(0, 10000, 2))  # lista grande
alvo = 5000

start = time.time()
idx_linear = busca_linear(lista, alvo)
end = time.time()
print(f"Linear: {idx_linear}, tempo: {end - start}")

start = time.time()
idx_binaria = busca_binaria(lista, alvo)
end = time.time()
print(f"Binária: {idx_binaria}, tempo: {end - start}")

File: cpp/gabaritos/gabarito26.cpp
#include <iostream>
#include <vector>
#include <chrono>

int busca_linear(const std::vector<int>& lst, int alvo) {
    for (size_t i = 0; i < lst.size(); ++i) {
        if (lst[i] == alvo) return i;
    }
    return -1;
}

int busca_binaria(const std::vector<int>& lst, int alvo) {
    int left = 0, right = lst.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (lst[mid] == alvo) return mid;
        else if (lst[mid] < alvo) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

int main() {
    std::vector<int> lista;
    for (int i = 0; i < 10000; i += 2) lista.push_back(i);
    int alvo = 5000;

    auto start = std::chrono::high_resolution_clock::now();
    int idx_linear = busca_linear(lista, alvo);
    auto end = std::chrono::high_resolution_clock::now();
    auto duration_linear = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    std::cout << "Linear: " << idx_linear << ", tempo: " << duration_linear.count() << " us" << std::endl;

    start = std::chrono::high_resolution_clock::now();
    int idx_binaria = busca_binaria(lista, alvo);
    end = std::chrono::high_resolution_clock::now();
    auto duration_binaria = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    std::cout << "Binária: " << idx_binaria << ", tempo: " << duration_binaria.count() << " us" << std::endl;

    return 0;
}